core/src/main/scala/org/bitcoins/core/protocol/tlv/TLV.scala

```diff
@@ -0,0 +1,142 @@
+package org.bitcoins.core.protocol.tlv
+
+import org.bitcoins.core.number.UInt16
+import org.bitcoins.core.protocol.BigSizeUInt
+import org.bitcoins.crypto.{Factory, NetworkElement}
+import scodec.bits.ByteVector
+
+sealed trait TLV extends NetworkElement {
+  def tpe: BigSizeUInt
+  def value: ByteVector
+
+  def length: BigSizeUInt = {
+    BigSizeUInt.calcFor(value)
+  }
+
+  override def bytes: ByteVector = {
+    tpe.bytes ++ length.bytes ++ value
+  }
+}
+
+object TLV extends Factory[TLV] {
+
+  def decodeTLV(bytes: ByteVector): (BigSizeUInt, BigSizeUInt, ByteVector) = {
```

Comment:
I would say we should just wrap this in a return type with a nice name and fields

I.e. `DecodeTLVResult` or something like that.